### PR TITLE
ListQueryObjectType should be filtered from the database

### PR DESCRIPTION
**Pull request by hyzx86**
Related to this issue https://github.com/OrchardCMS/OrchardCore/issues/15731, 
Not a fix, just optimizing query performance

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs [hunk 1]
```diff
@@ -35,15 +35,14 @@ public ListQueryObjectType(IStringLocalizer<ListQueryObjectType> S)
                     var session = serviceProvider.GetService<ISession>();
                     var accessor = serviceProvider.GetRequiredService<IDataLoaderContextAccessor>();
 
-                    var dataLoader = accessor.Context.GetOrAddCollectionBatchLoader<string, ContentItem>("ContainedPublishedContentItems", x => LoadPublishedContentItemsForListAsync(x, session));
+                    var dataLoader = accessor.Context.GetOrAddCollectionBatchLoader<string, ContentItem>("ContainedPublishedContentItems",
+                                                    x => LoadPublishedContentItemsForListAsync(x, session, g.GetArgument<int>("skip"), g.GetArgument<int>("first")));
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs:39]:
```suggestion
                    var dataLoader = accessor.Context.GetOrAddCollectionBatchLoader<string, ContentItem>("ContainedPublishedContentItems",
                       x => LoadPublishedContentItemsForListAsync(x, session, g.GetArgument<int>("skip"), g.GetArgument<int>("first")));
```

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs:39]: Does "first" specify the amount of the fetched rows?

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs (hunk 1 above)
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs]: Yes, in the context of the code, it defaults to 10

**Comment by hyzx86** [comment]
I'm trying to check other content queries, but it doesn't look like there's any more optimization left

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs [hunk 2]
```diff
@@ -35,15 +35,14 @@ public ListQueryObjectType(IStringLocalizer<ListQueryObjectType> S)
...
 
-                    return ((await dataLoader.LoadAsync(g.Source.ContentItem.ContentItemId).GetResultAsync())
-                                .Skip(g.GetArgument<int>("skip"))
-                                .Take(g.GetArgument<int>("first")));
+                    return await dataLoader.LoadAsync(g.Source.ContentItem.ContentItemId).GetResultAsync();
                 });
         }
 
-        private static async Task<ILookup<string, ContentItem>> LoadPublishedContentItemsForListAsync(IEnumerable<string> contentItemIds, ISession session)
+        private static async Task<ILookup<string, ContentItem>> LoadPublishedContentItemsForListAsync(IEnumerable<string> contentItemIds, ISession session, int skip, int first)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs:45]:
```suggestion
        private static async Task<ILookup<string, ContentItem>> LoadPublishedContentItemsForListAsync(IEnumerable<string> contentItemIds, ISession session, int skip, int count)
```

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs [hunk 3]
```diff
@@ -54,6 +53,8 @@ public ListQueryObjectType(IStringLocalizer<ListQueryObjectType> S)
                                      .With<ContentItemIndex>(ci => ci.Published)
                                      .With<ContainedPartIndex>(cp => cp.ListContentItemId.IsIn(contentItemIds))
                                      .OrderBy(o => o.Order)
+                                     .Skip(skip)
+                                     .Take(first)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs:57]:
```suggestion
                                     .Take(count)
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs (hunk 2 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs:45]: Now much better

**Review by hishamco — APPROVED** [review]

**Comment by Piedone** [comment]
Is there anything else before this can be merged, @hishamco? Please just merge if you're OK with it, to prevent any possible merge conflicts.

**Comment by hishamco** [comment]
I update to main then I will merge

**Pull request merged**